### ISSUE TITLE
Buildings without height still need to extrude starting at zoom 17/18

### ISSUE THREAD

**Issue by nvkelso** [tangrams/eraser-map#109]
This is important for routing.

This seems to work with WebGL, but is broken in the ES version in Eraser Map.

@blair1618, @bcamper, @meetar

**Comment by matteblair** [comment]
I see that buildings with extrusion are defined [here](https://github.com/tangrams/eraser-map/blob/gh-pages/eraser-map.yaml#L2229), and you have `extrude: true` in the two rules there. According to [the docs](https://github.com/tangrams/tangram-docs/blob/gh-pages/pages/draw.md#extrude), `extrude: true` should extrude buildings according to their `height` and `min_height` properties, so a feature without height should not be extruded by this rule. 

If those buildings are being extruded in WebGL then that would be a bug. 

Another possibility is that the map isn't zoomed enough in routing mode for buildings to match the `$zoom: { min: 18 }` filter, and therefore you aren't seeing as many buildings as you expect.

**Comment by nvkelso** [comment]
See also: https://github.com/mapzen/eraser-map/issues/166 which suggests the current max_zoom on the map in the Android app is 18 or 17.9 so I could see the `{ min: 18 }` filter being an issue. 

I think @meetar wrote this style originally, can you weigh in with suggested fix / WebGL versus ES comparison?

**Comment by bcamper** [comment]
Eh, it turns out (this is very very old code) that `extrude` has a default value of 20 meters, if the `height` is not defined:

https://github.com/tangrams/tangram/blob/master/src/styles/polygons/polygons.js#L59
https://github.com/tangrams/tangram/blob/master/src/styles/style_parser.js#L122

We could deal with this by:
- Changing this behavior, and adjusting any previous stylesheets as necessary.
  - If we do this, we could add an `extrude_default` (similar to `sprite_default`), so it is configurable instead of hard-coded. Though this can also be accomplished in less elegant fashion with `extrude: function() { return feature.height || 20; }`.
- Implementing the default in ES.
- Other?

Sorry, this is kind of a weird once since it goes way back to when I first added extrusion :)  And I guess there is some argument that if you put `extrude: true` on a feature, it's helpful to get the immediate feedback by seeing it raised. Unsure what is most desirable here...

**Comment by nvkelso** [comment]
This feels similar to the default [text font size of 12](https://github.com/tangrams/tangram-docs/blob/gh-pages/pages/draw.md#size-1)
…

If I extrude a building I expect it to extrude to some default. **I vote
for implementing the default in ES.**

And in the meantime we could hack it for Eraser Map scene with `extrude:
function() { return feature.height || 20; }`, as Brett suggests.

**Comment by bcamper** [comment]
I do think it's friendlier from a learning-the-system perspective, type
less to get a result (the counter argument is that defaults can be
confusing).

**Comment by matteblair** [comment]
I don't think this default extrusion behavior is a good idea. 

It's easy to accomplish a default extrusion height in a style; ~~a one-line position injection will do splendidly for the eraser-map style~~ the JS function mentioned above is a pretty intuitive solution. 

Adding implicit, exceptional behavior like this makes it more difficult to reason about the behavior of a system that's already quite complex.

**Comment by nvkelso** [comment]
@bcamper and @blair1618 I suspect this needs to be set in the shader itself and not the basic YAML extrude property. That is to say, setting `extrude: function() { return feature.height || 20; }` makes buildings with height infinitely tall on device, and buildings without height still aren't extruded.

```
$ git diff
diff --git a/eraser-map.yaml b/eraser-map.yaml
index 8fc5379..e80087e 100644
--- a/eraser-map.yaml
+++ b/eraser-map.yaml
@@ -2292,20 +2292,12 @@ layers:
                 polygons:
                     visible: true
                     style: building-grid
-                    extrude: function() { return feature.height || 20; }
+                    extrude: true
                     color: [.885, .870, .870]
                 lines:
                     visible: true
                     style: building-lines
-                    extrude: function() { return feature.height || 20; }
-                    
-            buildings-z17-boost:
-                filter: { $zoom: { min: 17, max: 18 }, height: false }
-                draw:
-                    polygons:
-                        extrude: 30
-                    lines:
-                        extrude: 30
+                    extrude: true
```

 I wonder if the height is controlled with shader magic here:

```
    scale-buildings:
        shaders:
            blocks:
                position: |
                    // scale buildings based on zoom
                    float zoom = u_map_position.z;
                    float min = .1;         // minimum building scale
                    float midpoint = 16.; // middle of zoom range
                    float speed = 2.;  // number of zoom levels to transition over
                    float e = (zoom - midpoint) / (speed * .2);
                    position.z *= ((1. - min) / (1. + (exp(e)))) + min;
```

Can @bcamper take a look at this? I don't have enough experience to set this, and it's hard also to test and not gum up the main Android build.

**Comment by nvkelso** [comment]
@sensescape Can you attach the awesome screenshot here?

**Comment by sensescape** [comment]
![screenshot_2015-12-07-14-00-26](https://cloud.githubusercontent.com/assets/466585/11641063/ff60383a-9cea-11e5-858c-4de848a9aca8.png)

**Comment by bcamper** [comment]
Sounds like a feature.

**Comment by matteblair** [comment]
Yeah, I think I've just seen what you're talking about :)

This should be a correct way to write this behavior in the stylesheet. I'm unclear why it's totally bonkers on ES right now but I have two theories:
1. JS functions may not be enabled for "extrude". I thought this was enabled already, but if not it's quite simple to enable JS functions on a parameter.
2. Numeric values for "extrude" might not be interpreted in meters. I'm pretty confident this is true, and it'll be fixed with one multiplication. 

Thanks for giving it a spin; hang tight while I un-crazy those extrusions!

**Comment by matteblair** [comment]
Progress: it turns out functions _do_ work for 'extrude' but only if you return [min_height, height], so I'll add a case to allow returning a single value (like on JS). The unit values are also scaled incorrectly, like I thought, and that's a simple change.

**Comment by nvkelso** [comment]
Fixed in d0ad47eb9b1784b215251128eb7000ccf4dcdb5d and verified in master-569 on device!
